libs/ng-mocks/src/lib/common/ng-mocks-stack.ts

```diff
@@ -138,11 +138,30 @@ const installJestCircus = () => {
   return true;
 };
 
+// istanbul ignore next
+const installMochaReporter = () => {
+  try {
+    mocha.setup({
+      rootHooks: {
+        afterAll: stackPop,
+        beforeAll: () => stackPush('root'),
```

Comment:
Can it be that beforeAll is used several times per test file?

If it is so, then it should be `suite`.

`root` is provided by `ng-mocks` on the start.

Reply:
According to https://mochajs.org/#available-root-hooks, the `beforeAll`/`afterAll` root hooks behave differently depending on if mocha is run in serial or parallel mode. They seem mostly synonymous to ng-mocks `root`, but in parallel mode mocha has multiple "roots". I believe parallel mode would instantiate a different ng-mocks instance per each parallel run, so it should be fine as `root`, but I cannot verify as we use serial mode.

Reply:
Can you do something like that with mocha:

```
describe
  beforeAll
  afterAll
  it
  it
  it
  describe
    beforeAll
    afterAll
    it
    it
```

so the first beforeAll covers the nested describe too, whereas the second beforeAll covers only its suite.

Reply:
Yes you can, but the hooks here do not cause each `beforeAll`/`afterAll` to behave differently, they just register a single `beforeAll`/`afterAll` at the root of the test suite.

So even if you had what you described, the root hooks would be like so:
```
beforeAll # registered by root hook
afterAll # registered by root hook
describe
  beforeAll
  afterAll
  it
  it
  it
  describe
    beforeAll
    afterAll
    it
    it
```

Reply:
Yep..... now I'm wondering how it works if I used `MockInstance` in the nested `beforeAll`....

Reply:
I believe it would still work fine. The `afterAll` would still make sure it gets cleaned up

Reply:
in a case of manual reset - yes. But if I customized a service in the first beforeAll and then customized it again in the nested beforeAll, then this change should disappear once the describe has been finished:

```
beforeAll # registered by root hook
afterAll # registered by root hook
describe
  beforeAll
  afterAll
  it -- 1st customization
  describe
    beforeAll
    afterAll
    it -- 2nd customization
  it -- 2nd is removed and 1st is present
```

Reply:
Anyway it doesn't affect your PR and adds one more point to the solution with a manual control of the stack state which will be implemented as a separate PR.